### PR TITLE
Sorr task918 add unit test for split universe

### PR DESCRIPTION
**Pull request by smitpatel49**
#918

### CODE REVIEW

File: im_v2/common/data/extract/test/test_extract_utils.py [hunk 1]
```diff
@@ -826,6 +826,62 @@ def test_download_and_resample_bid_ask_data(self) -> None:
...
+        "INJ_USDT",
+        "NEAR_USDT",
+    ]
+
+    def test1(self) -> None:
+        """
+        Test if both group_size and universe_part are 0.
+        """
+        universe_subset = imvcdeexut._split_universe(self.universe, 0, 0)
+        actual_str = " ".join([str(element) for element in universe_subset])
```
Comment (sonaalKant) [inline im_v2/common/data/extract/test/test_extract_utils.py:845]: Lets not do this, instead convert the output from the method to str and then compare directly.

**Review by sonaalKant — CHANGES_REQUESTED** [review]
I will also add a corner case where group is 4 and part is 2.

File: im_v2/common/data/extract/test/test_extract_utils.py (hunk 1 above)
Comment (smitpatel49) [reply im_v2/common/data/extract/test/test_extract_utils.py:845]: Hi, I will add that corner case. The output from the method is of list type, I felt that this was efficient to convert it to string type. Can you suggest a more suitable approach? Thanks.

File: im_v2/common/data/extract/test/test_extract_utils.py [hunk 2]
```diff
@@ -820,6 +819,77 @@ def test_download_and_resample_bid_ask_data(self) -> None:
...
+            map(str, imvcdeexut._split_universe(self.universe, 6, 1))
+        )
+        expected_str = r"""
+        ALICE_USDT GALA_USDT FLOW_USDT HBAR_USDT INJ_USDT NEAR_USDT
+        """
+        self.assert_equal(actual_str, expected_str, fuzzy_match=True)
+
+    def test3(self) -> None:
+        """
+        Test if roup_size is 6 and universe_part is 2.
```
Comment (gpsaggese) [inline im_v2/common/data/extract/test/test_extract_utils.py:861]: -> group_size

File: im_v2/common/data/extract/test/test_extract_utils.py [hunk 3]
```diff
@@ -820,6 +819,77 @@ def test_download_and_resample_bid_ask_data(self) -> None:
...
+        "INJ_USDT",
+        "NEAR_USDT",
+    ]
+
+    def test1(self) -> None:
+        """
+        Test if both group_size and universe_part are 0.
+        """
+        actual_str = " ".join(
+            map(str, imvcdeexut._split_universe(self.universe, 0, 0))
```
Comment (gpsaggese) [inline im_v2/common/data/extract/test/test_extract_utils.py:838]: Pls read our doc with attention (e.g., docs/coding/all.write_unit_tests.how_to_guide.md and more)

Let's rewrite as
```
# Prepare inputs.
group_size = 0
universe_part = 0
# Call function under test.
act = imvcdeexut._split_universe(self.universe, group_size, universe_part)
# Check the output.
actual_str = " ".join(map(str, ...)
```

File: im_v2/common/data/extract/test/test_extract_utils.py [hunk 4]
```diff
@@ -820,6 +819,77 @@ def test_download_and_resample_bid_ask_data(self) -> None:
...
+        """
+        self.assert_equal(actual_str, expected_str, fuzzy_match=True)
+
+    def test2(self) -> None:
+        """
+        Test if group_size is 6 and universe_part is 1.
+
+        (This returns the whole list in the 1st part)
+        """
+        actual_str = " ".join(
```
Comment (gpsaggese) [inline im_v2/common/data/extract/test/test_extract_utils.py:851]: Let's factor out all the common code in a function and then you pass the input and the expected_str.

File: im_v2/common/data/extract/test/test_extract_utils.py [hunk 5]
```diff
@@ -820,6 +819,77 @@ def test_download_and_resample_bid_ask_data(self) -> None:
...
+        expected_str = r"""
+
+        """
+        self.assert_equal(actual_str, expected_str, fuzzy_match=True)
+
+    def test4(self) -> None:
+        """
+        Test to check for error.
+        """
+        with pytest.raises(RuntimeError):
```
Comment (gpsaggese) [inline im_v2/common/data/extract/test/test_extract_utils.py:877]: docs/coding/all.write_unit_tests.how_to_guide.md:31:    + [Always explain `self.assertRaises`](#always-explain-selfassertraises

**Review by gpsaggese — COMMENTED** [review]
Going in the right direction, but you need to read our docs and our code base more carefully.

File: im_v2/common/data/extract/test/test_extract_utils.py (hunk 4 above)
Comment (smitpatel49) [reply im_v2/common/data/extract/test/test_extract_utils.py:851]: I understand the changes requested and will make them as you mentioned. I will go through the documentation once again. Thanks for the review @gpsaggese.

File: im_v2/common/data/extract/test/test_extract_utils.py (hunk 5 above)
Comment (smitpatel49) [reply im_v2/common/data/extract/test/test_extract_utils.py:877]: I will put up the proper explanation. However, I cannot find `self.assertRaises` in the documentation.

**Comment by smitpatel49** [comment]
I am sorry for another PR. I made a minor mistake, but the code is not changed. I would request you to review it @gpsaggese, @sonaalKant, @samarth9008.

File: im_v2/common/data/extract/test/test_extract_utils.py [hunk 6]
```diff
@@ -820,6 +819,92 @@ def test_download_and_resample_bid_ask_data(self) -> None:
         self.check_resampler()
 
 
+class TestSplitUniverse(hunitest.TestCase):
```
Comment (samarth9008) [inline im_v2/common/data/extract/test/test_extract_utils.py:822]: Lets add class doc string

```
Test that `_split_universe()` works correctly.
```

File: im_v2/common/data/extract/test/test_extract_utils.py [hunk 7]
```diff
@@ -820,6 +819,92 @@ def test_download_and_resample_bid_ask_data(self) -> None:
...
+    # Prepare inputs.
+    universe = [
+        "ALICE_USDT",
+        "GALA_USDT",
+        "FLOW_USDT",
+        "HBAR_USDT",
+        "INJ_USDT",
+        "NEAR_USDT",
+    ]
```
Comment (samarth9008) [inline im_v2/common/data/extract/test/test_extract_utils.py:831]: I would move this inside the `get_universe_part()` function since it is constant for all the test cases

File: im_v2/common/data/extract/test/test_extract_utils.py [hunk 8]
```diff
@@ -820,6 +819,92 @@ def test_download_and_resample_bid_ask_data(self) -> None:
...
+    group_size = 0
+    universe_part = 0
```
Comment (samarth9008) [inline im_v2/common/data/extract/test/test_extract_utils.py:833]: Add these two as params to `get_universe_part()` so you can test them with different values

File: im_v2/common/data/extract/test/test_extract_utils.py [hunk 9]
```diff
@@ -820,6 +819,92 @@ def test_download_and_resample_bid_ask_data(self) -> None:
...
+
+    # Call function under test.
```
Comment (samarth9008) [inline im_v2/common/data/extract/test/test_extract_utils.py:835]: No comment above the function

Comment (samarth9008) [reply im_v2/common/data/extract/test/test_extract_utils.py:835]: Instead there should be a doc string for the function

File: im_v2/common/data/extract/test/test_extract_utils.py [hunk 10]
```diff
@@ -820,6 +819,92 @@ def test_download_and_resample_bid_ask_data(self) -> None:
...
+    def get_universe_part(self) -> List:
```
Comment (samarth9008) [inline im_v2/common/data/extract/test/test_extract_utils.py:836]: make sure to add doc string of the function accordingly

Comment (samarth9008) [reply im_v2/common/data/extract/test/test_extract_utils.py:836]: Lets also rename `get_universe_part` -> `helper`

File: im_v2/common/data/extract/test/test_extract_utils.py [hunk 11]
```diff
@@ -820,6 +819,92 @@ def test_download_and_resample_bid_ask_data(self) -> None:
...
+        actual_output = imvcdeexut._split_universe(
+            self.universe, self.group_size, self.universe_part
+        )
+        # Return the actual output.
```
Comment (samarth9008) [inline im_v2/common/data/extract/test/test_extract_utils.py:840]: No need of this as we can read it from the code. Comments are for why the code is doing this rather than what the code is doing.

File: im_v2/common/data/extract/test/test_extract_utils.py [hunk 12]
```diff
@@ -820,6 +819,92 @@ def test_download_and_resample_bid_ask_data(self) -> None:
...
+        return actual_output
+
+    def test1(self) -> None:
+        """
+        Test if both group_size and universe_part are 0.
```
Comment (samarth9008) [inline im_v2/common/data/extract/test/test_extract_utils.py:845]: param name in comment should be surrounded by back ticks.

File: im_v2/common/data/extract/test/test_extract_utils.py [hunk 13]
```diff
@@ -820,6 +819,92 @@ def test_download_and_resample_bid_ask_data(self) -> None:
...
+        """
+        actual_str = str(self.get_universe_part())
```
Comment (samarth9008) [inline im_v2/common/data/extract/test/test_extract_utils.py:847]: I would do something like

```
actual = self.get_universe_part()
actual = str(actual)
expected = r"""
        []
        """
```

make sure this pattern is reflected everywhere

File: im_v2/common/data/extract/test/test_extract_utils.py [hunk 14]
```diff
@@ -820,6 +819,92 @@ def test_download_and_resample_bid_ask_data(self) -> None:
...
+        expected_str = r"""
+        []
+        """
+        self.assert_equal(actual_str, expected_str, fuzzy_match=True)
+
+    def test2(self) -> None:
+        """
+        Test if group_size is 6 and universe_part is 1.
+
+        (This returns the whole list in the 1st part)
```
Comment (samarth9008) [inline im_v2/common/data/extract/test/test_extract_utils.py:857]: No need for this kind of comment in testing. Remove everywhere if present

File: im_v2/common/data/extract/test/test_extract_utils.py (hunk 8 above)
Comment (samarth9008) [reply im_v2/common/data/extract/test/test_extract_utils.py:833]: Make sure to change the other part of code accordingly

File: im_v2/common/data/extract/test/test_extract_utils.py [hunk 15]
```diff
@@ -820,6 +819,92 @@ def test_download_and_resample_bid_ask_data(self) -> None:
...
+        self.universe_part = 2
+        actual_str = str(self.get_universe_part())
+        expected_str = r"""
+        ['INJ_USDT', 'NEAR_USDT']
+        """
+        self.assert_equal(actual_str, expected_str, fuzzy_match=True)
+
+    def test5(self) -> None:
+        """
+        Test to check for error.
```
Comment (samarth9008) [inline im_v2/common/data/extract/test/test_extract_utils.py:895]: -> Test if the error is raised if no such part exists.

**Review by samarth9008 — COMMENTED** [review]
Some nits.

**Comment by smitpatel49** [comment]
Thanks for the review @samarth9008, I have updated the code accordingly.

**Comment by samarth9008** [comment]
Pls resolve the conversation if the changes are made.

File: im_v2/common/data/extract/test/test_extract_utils.py [hunk 16]
```diff
@@ -820,6 +819,81 @@ def test_download_and_resample_bid_ask_data(self) -> None:
...
+        ]
+        actual_output = imvcdeexut._split_universe(
+            universe, group_size, universe_part
+        )
+        return actual_output
+
+    def test1(self) -> None:
+        """
+        Test if both `group_size` and `universe_part` are 0.
+        """
```
Comment (DanilYachmenev) [inline im_v2/common/data/extract/test/test_extract_utils.py:847]: In very rare cases we can refer to objects on docstrings, unit test methods are such cases sometimes
But it should be demonstrated rather like a list of param-value elements in order to ease processing for readers
since you already have class docstring, here i'd do simply
```
"""
group_size = 0
universe_part = 0
"""
```

File: im_v2/common/data/extract/test/test_extract_utils.py [hunk 17]
```diff
@@ -820,6 +819,81 @@ def test_download_and_resample_bid_ask_data(self) -> None:
...
+        actual_str = str(self.helper(0, 0))
+        expected_str = r"""
+                    []
+                    """
```
Comment (DanilYachmenev) [inline im_v2/common/data/extract/test/test_extract_utils.py:851]: unnecessary indentations

should be 
```
expected_str = r"""
[]
"""
```
you've run linter, right?
weird it hasn't fixed it

File: im_v2/common/data/extract/test/test_extract_utils.py [hunk 18]
```diff
@@ -820,6 +819,81 @@ def test_download_and_resample_bid_ask_data(self) -> None:
...
+
+    def test2(self) -> None:
+        """
+        Test if `group_size` is 6 and `universe_part` is 1.
+        """
+        actual_str = str(self.helper(6, 1))
+        expected_str = r"""
+                    ['ALICE_USDT', 'GALA_USDT', 'FLOW_USDT', 'HBAR_USDT', 'INJ_USDT', 'NEAR_USDT']
+                    """
+        self.assert_equal(actual_str, expected_str, fuzzy_match=True)
```
Comment (DanilYachmenev) [inline im_v2/common/data/extract/test/test_extract_utils.py:862]: same comments for all the other test methods

File: im_v2/common/data/extract/test/test_extract_utils.py [hunk 19]
```diff
@@ -820,6 +819,81 @@ def test_download_and_resample_bid_ask_data(self) -> None:
...
+        actual_str = str(self.helper(4, 2))
+        expected_str = r"""
+                    ['INJ_USDT', 'NEAR_USDT']
+                    """
+        self.assert_equal(actual_str, expected_str, fuzzy_match=True)
+
+    def test5(self) -> None:
+        """
+        Test if the error is raised if no such part exists.
+        """
```
Comment (DanilYachmenev) [inline im_v2/common/data/extract/test/test_extract_utils.py:887]: here we have an assert so
```
"""
Check that the error is raised if universe does not have input part.

group_size = 6
universe_part = 6
"""
```

I'd even take `universe_part = 7` just for more clarity

**Comment by DanilYachmenev** [comment]
pls read https://github.com/kaizen-ai/kaizenflow/blob/master/docs/coding/all.submit_code_for_review.how_to_guide.md#save-reviewer-time, particularly part about managing PR labels
pls follow it

File: im_v2/common/data/extract/test/test_extract_utils.py [hunk 20]
```diff
@@ -1746,230 +1820,3 @@ def test_resample_rt_bid_ask_data_periodically(
             1 1709281440000 3384.71 1.617 3384.72 109.136 3384.715 0.005 -4.212022 3384.4 34.431 3384.41 48.920 3384.405 0.005 -0.351229 3384.71 34.431 3384.72 129.577 3384.715 0.005 -0.351229 3384.4 0.610 3384.41 48.920 3384.405 0.005 -5.358572 3384.667500 3.106187 3384.677500 104.250812 3384.672500 0.005 -4.484592 0.0579 0.0 343.623874 323.847900 binance ETH_USDT 1 2024-02-20 18:01:01+00:00
             """
             self.check_df_output(resampled_df2, None, None, None, exp_df2)
```
Comment (gpsaggese) [inline im_v2/common/data/extract/test/test_extract_utils.py:1822]: What happened to this deleted code? Is it a missing merge from master?

Comment (smitpatel49) [reply im_v2/common/data/extract/test/test_extract_utils.py:1822]: I think it might be a merge missing from master, as I am sure I haven't changed any other part of the code.

File: im_v2/common/data/extract/test/test_extract_utils.py (hunk 17 above)
Comment (smitpatel49) [reply im_v2/common/data/extract/test/test_extract_utils.py:851]: I did. I made the change later when I was confused from another review. I undid it again.

File: im_v2/common/data/extract/test/test_extract_utils.py (hunk 20 above)
Comment (DanilYachmenev) [reply im_v2/common/data/extract/test/test_extract_utils.py:1859]: @smitpatel49 pls return this code back since it is still in master

File: im_v2/common/data/extract/test/test_extract_utils.py [hunk 21]
```diff
@@ -820,6 +819,118 @@ def test_download_and_resample_bid_ask_data(self) -> None:
...
+
+        group_size = 6 universe_part = 6
+        """
+        # Raise an error for no input part.
+        with self.assertRaises(RuntimeError) as cm:
+            self.helper(6, 6)
+        # Get the error output from helper function.
+        actual = cm.exception
+        # Check the output.
+        actual_str = str(actual)
```
Comment (DanilYachmenev) [inline im_v2/common/data/extract/test/test_extract_utils.py:909]: merge it all in one line
no need for comments, self-explanatory lines
```
actual = str(cm.exception)
```

File: im_v2/common/data/extract/test/test_extract_utils.py [hunk 22]
```diff
@@ -820,6 +819,118 @@ def test_download_and_resample_bid_ask_data(self) -> None:
...
+        """
+        universe = [
+            "ALICE_USDT",
+            "GALA_USDT",
+            "FLOW_USDT",
+            "HBAR_USDT",
+            "INJ_USDT",
+            "NEAR_USDT",
+        ]
+        # Call function under test.
```
Comment (DanilYachmenev) [inline im_v2/common/data/extract/test/test_extract_utils.py:839]: no need as well, self-explanatory moment

we aim to comment as much as possible, but in some cases, mostly in tests, such comments are not really needed since tests are all about calling a tested function and comparing its outcome with expected

what you need to comment are some pivotal code moments which are not self-evident from the action athat is being implemented

File: im_v2/common/data/extract/test/test_extract_utils.py [hunk 23]
```diff
@@ -820,6 +819,118 @@ def test_download_and_resample_bid_ask_data(self) -> None:
...
+        actual_output = imvcdeexut._split_universe(
+            universe, group_size, universe_part
+        )
+        return actual_output
+
+    def test1(self) -> None:
+        """
+        group_size = 0 universe_part = 0.
```
Comment (DanilYachmenev) [inline im_v2/common/data/extract/test/test_extract_utils.py:847]: split on 2 lines
```
"""
group_size = 0
universe_part = 0
"""
```
or does linter fixes it?

if the poblem is i linter, lets move docstring line from the class docstring to the each test method (except for the one with error exception check) and remove it from the class
so it will be

```
"""
Check that universe is split correctly.

group_size = 0
universe_part = 0
"""
```

File: im_v2/common/data/extract/test/test_extract_utils.py [hunk 24]
```diff
@@ -820,6 +819,118 @@ def test_download_and_resample_bid_ask_data(self) -> None:
...
+        """
+        # Get the actual output from helper function.
```
Comment (DanilYachmenev) [inline im_v2/common/data/extract/test/test_extract_utils.py:849]: no need for this comment, helper docstring explain it

File: im_v2/common/data/extract/test/test_extract_utils.py [hunk 25]
```diff
@@ -820,6 +819,118 @@ def test_download_and_resample_bid_ask_data(self) -> None:
...
+        actual = self.helper(0, 0)
+        # Check the output.
+        actual_str = str(actual)
```
Comment (DanilYachmenev) [inline im_v2/common/data/extract/test/test_extract_utils.py:852]: you do this in every test method
so lets move it to helper function and change its docstring to
`Run the function and get a string representation of its output.`

File: im_v2/common/data/extract/test/test_extract_utils.py [hunk 26]
```diff
@@ -820,6 +819,118 @@ def test_download_and_resample_bid_ask_data(self) -> None:
...
+        """
+        self.assert_equal(actual_str, expected_str, fuzzy_match=True)
+
+    def test2(self) -> None:
+        """
+        group_size = 6 universe_part = 1.
+        """
+        # Get the actual output from helper function.
+        actual = self.helper(6, 1)
+        # Check the output.
```
Comment (DanilYachmenev) [inline im_v2/common/data/extract/test/test_extract_utils.py:864]: overall no need for comments here
they repeat over and over without doing much really

although you can add clarity this way:
```
group_size = 6
universe_part = 1
actual = self.helper(group_size, universe_part)
expected = r"""
...
"""
self.assert_equal(actual, expected)
```

fix all the other test methods accordingly
read https://github.com/kaizen-ai/kaizenflow/blob/master/docs/coding/all.coding_style.how_to_guide.md#dont-hardwire-params-in-a-function-call

File: im_v2/common/data/extract/test/test_extract_utils.py (hunk 20 above)
Comment (smitpatel49) [reply im_v2/common/data/extract/test/test_extract_utils.py:1859]: I have added the code back. I am still uncertain of how it got deleted. Sorry for the inconvenience.

File: im_v2/common/data/extract/test/test_extract_utils.py [hunk 27]
```diff
@@ -820,6 +819,106 @@ def test_download_and_resample_bid_ask_data(self) -> None:
         self.check_resampler()
 
 
+class TestSplitUniverse(hunitest.TestCase):
+    def helper(self, group_size: int, universe_part: int) -> List:
+        """
+        Run the function and get ta string representation of its output.
```
Comment (DanilYachmenev) [inline im_v2/common/data/extract/test/test_extract_utils.py:825]: -> `... and get a string ...`
typo

File: im_v2/common/data/extract/test/test_extract_utils.py (hunk 25 above)
Comment (DanilYachmenev) [reply im_v2/common/data/extract/test/test_extract_utils.py:847]: this one is not done
move `actual_str = str(actual)` to `helper()` function and update each test method accordingly

File: im_v2/common/data/extract/test/test_extract_utils.py [hunk 28]
```diff
@@ -820,6 +819,106 @@ def test_download_and_resample_bid_ask_data(self) -> None:
...
+        Check that universe is split correctly.
+        """
+        group_size = 0
+        universe_part = 0
+        actual = self.helper(group_size, universe_part)
+        actual_str = str(actual)
+        expected_str = r"""
+        []
+        """
+        self.assert_equal(actual_str, expected_str, fuzzy_match=True)
```
Comment (DanilYachmenev) [inline im_v2/common/data/extract/test/test_extract_utils.py:851]: do we need `fuzzy_match=True`?
try to run without it and if it is required lets keep it but if not - remove

File: im_v2/common/data/extract/test/test_extract_utils.py [hunk 29]
```diff
@@ -820,6 +819,106 @@ def test_download_and_resample_bid_ask_data(self) -> None:
...
+        group_size = 6
+        universe_part = 6
+        # Raise an error for no input part.
+        with self.assertRaises(RuntimeError) as cm:
+            self.helper(group_size, universe_part)
+        actual_str = str(cm.exception)
+        expected_str = r"""
+        Universe does not have 6 parts of 6 pairs.    It has 6 symbols.
+        """
+        self.assert_equal(actual_str, expected_str, fuzzy_match=True)
```
Comment (DanilYachmenev) [inline im_v2/common/data/extract/test/test_extract_utils.py:905]: this one is basically identical to `test6`, no need, remove

File: im_v2/common/data/extract/test/test_extract_utils.py [hunk 30]
```diff
@@ -820,6 +819,106 @@ def test_download_and_resample_bid_ask_data(self) -> None:
...
+
+    def test6(self) -> None:
```
Comment (DanilYachmenev) [inline im_v2/common/data/extract/test/test_extract_utils.py:907]: remember to name it `test5` after you implement the comment above

**Review by DanilYachmenev — COMMENTED** [review]
fix and ready to merge

File: im_v2/common/data/extract/test/test_extract_utils.py (hunk 28 above)
Comment (smitpatel49) [reply im_v2/common/data/extract/test/test_extract_utils.py:851]: I think it is required as I tried without it and the `expected_str` to be in one line.

**Review by DanilYachmenev — COMMENTED** [review]
lgtm

**Review by samarth9008 — COMMENTED** [review]
LG

**Review by samarth9008 — APPROVED** [review]
